sycl-fusion/common/include/Kernel.h

```diff
@@ -0,0 +1,155 @@
+
+  // Explicit constructor for compatibility with LLVM YAML I/O.
+  SYCLKernelAttribute() : Values{} {};
+  SYCLKernelAttribute(std::string Name)
+      : AttributeName{std::move(Name)}, Values{} {}
+
+  std::string AttributeName;
+  AttributeValueList Values;
+};
+
+enum ArgUsage : unsigned char {
+  // Used to indicate that an argument is not used by the kernel
+  Unused = 0,
+  // Used to indicate that an argument is used by the kernel
+  Used = 1u,
+  // Used to indicate that the accessor/pointer argument has been promoted to
+  // private memory
+  PromotedPrivate = 1u << 4,
+  // Used to indicate that the accessor/pointer argument has been promoted to
+  // local memory
+  PromotedLocal = 1u << 5,
+};
+
+///
+/// Encode usage of parameters for the actual kernel function.
+// This is a vector of unsigned char, because std::vector<bool> is a weird
+// construct and unlike all other std::vectors, and LLVM YAML I/O is having a
+// hard time coping with it.
+using ArgUsageMask = std::vector<std::underlying_type_t<ArgUsage>>;
+
+///
+/// Describe the list of arguments by their kind.
+struct SYCLArgumentDescriptor {
+
+  // Explicit constructor for compatibility with LLVM YAML I/O.
+  SYCLArgumentDescriptor() : Kinds{}, UsageMask{} {}
+
+  std::vector<ParameterKind> Kinds;
+
+  ArgUsageMask UsageMask;
+};
+
+///
+/// List of SYCL/OpenCL kernel attributes.
+using AttributeList = std::vector<SYCLKernelAttribute>;
+
+/// Information about a kernel from DPC++.
+struct SYCLKernelInfo {
+
+  std::string Name;
+
+  SYCLArgumentDescriptor Args;
+
+  AttributeList Attributes;
+
+  SYCLKernelBinaryInfo BinaryInfo;
+
+  //// Explicit constructor for compatibility with LLVM YAML I/O.
+  SYCLKernelInfo() : Name{}, Args{}, Attributes{}, BinaryInfo{} {}
+
+  SYCLKernelInfo(const std::string &KernelName,
+                 const SYCLArgumentDescriptor &ArgDesc,
+                 const SYCLKernelBinaryInfo &BinInfo)
+      : Name{KernelName}, Args{ArgDesc}, Attributes{}, BinaryInfo{BinInfo} {}
+
+  explicit SYCLKernelInfo(const std::string &KernelName)
+      : Name{KernelName}, Args{}, Attributes{}, BinaryInfo{} {}
+};
+
+///
+/// Represents a SPIR-V translation unit containing SYCL kernels by the
+/// KernelInfo for each of the contained kernels.
+class SYCLModuleInfo {
+public:
+  using KernelInfoList = std::vector<SYCLKernelInfo>;
+
+  void addKernel(SYCLKernelInfo &Kernel) { Kernels.push_back(Kernel); }
+
+  KernelInfoList &kernels() { return Kernels; }
+
+  bool hasKernelFor(const std::string &KernelName) {
+    return findKernelFor(KernelName) != nullptr;
+  }
+
+  SYCLKernelInfo *getKernelFor(const std::string &KernelName) {
+    return findKernelFor(KernelName);
+  }
+
+private:
+  SYCLKernelInfo *findKernelFor(const std::string &KernelName) {
+    auto It =
+        std::find_if(Kernels.begin(), Kernels.end(),
+                     [&](SYCLKernelInfo &K) { return K.Name == KernelName; });
+    return (It != Kernels.end()) ? &*It : nullptr;
+  }
```

Comment:
Unless I am missing something, it seems like the implementation of this function could be moved to `getKernelFor` and `hasKernelFor` could use that directly.

Reply:
You're right, thanks.